File: src/solar/solar-common/util/unit/UnitDeathTimeUtil.ts
/**
 * 单位死亡时间 记录工具
 * 主要用于单位handle 重用 清理等相关判断
 * 保证所有单位死亡或移除时一定要设置死亡时间。
 */
export default class UnitDeathTimeUtil {
    private static _sl_deathTimeData: { [hid: number]: number } = {}


    /**
     * 设置死亡时间 (毫秒单位)
     * @param deathUnit
     * @param deathTime
     */
    static setDeathTime(deathUnit: unit, deathTime: number = time) {
        UnitDeathTimeUtil._sl_deathTimeData[GetHandleId(deathUnit)] = deathTime;
    }

    /**
     * 不要依靠清除死亡时间。因为没有完善的清除逻辑(主要是英雄死亡后可复活 单位死亡后也可被复活技能复活)。只能允许新的重用单位 继承  旧死亡单位的死亡时间。但使用时保证单位是死亡的再使用就对了。
     */
    /**
     * 清除单位handle死亡时间 (以减少重用后的handle获得旧handle数据)
     * @param deathUnit
     * @param deathTime
     */
    static clearDeathTime(deathUnit: unit) {
        let hid = GetHandleId(deathUnit);
        if (UnitDeathTimeUtil._sl_deathTimeData[hid] || UnitDeathTimeUtil._sl_deathTimeData[hid] == 0) {
            UnitDeathTimeUtil._sl_deathTimeData[hid] = null;
        }
    }

    /**
     * 是否拥有死亡时间
     * @param deathUnit
     */
    static hasDeathTime(deathUnit: unit): boolean {
        if (UnitAlive(deathUnit)) {
            //还是活着的单位 没有死亡时间
            return false;
        }
        let d = UnitDeathTimeUtil._sl_deathTimeData[GetHandleId(deathUnit)];
        if (d == null) {
            return false;
        }
        return true;
    }

    /**
     * 获取死亡时间 (毫秒单位) 必须单位是死亡的 才能获取正确的死亡时间，活着的单位可能获取继承的旧单位死亡时间。
     * @param deathUnit
     */
    static getDeathTime(deathUnit: unit): number {
        return UnitDeathTimeUtil._sl_deathTimeData[GetHandleId(deathUnit)];
    }

    /**
     * 判断单位死亡是否已经过了（>=）指定时间 （ 必须对已经设置过死亡时间的单位进行判断
     * (在SolarDataClearState中已经对死亡单位时和移除单位时调用了setDeathTime来设置死亡时间)
     * @param deathUnit
     * @param timeSlot (秒单位)
     * @param   如未设定setDeathTime 则默认返回false 且报错
     */
    static isTimeOfDeathExceeded(deathUnit: unit, timeSlot: number = handleReuseMinTime) {
        if (UnitAlive(deathUnit)) {
            //还是活着的单位
            return false;
        }
        let deathTime = UnitDeathTimeUtil.getDeathTime(deathUnit);
        if (deathTime == null || deathTime == 0) {
            log.errorWithTraceBack("正在对一个已死亡但没有设置DeathTime的单位进行死亡时间判断！请先设置死亡单位的死亡时间")
            return false;
        }
        if ((time - deathTime) >= (timeSlot * 1000)) {
            return true;
        }
        return false;
    }


}

File: scripts/solar/solar-common/util/unit/UnitDeathTimeUtil.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["6"] = 6,["7"] = 6,["8"] = 6,["10"] = 6,["11"] = 15,["12"] = 15,["13"] = 15,["15"] = 16,["16"] = 15,["17"] = 27,["18"] = 28,["19"] = 29,["20"] = 30,["22"] = 27,["23"] = 38,["24"] = 39,["25"] = 41,["27"] = 43,["28"] = 44,["29"] = 45,["31"] = 47,["32"] = 38,["33"] = 54,["34"] = 55,["35"] = 54,["36"] = 65,["37"] = 65,["38"] = 65,["40"] = 66,["41"] = 68,["43"] = 70,["44"] = 71,["45"] = 72,["46"] = 73,["48"] = 75,["49"] = 76,["51"] = 78,["52"] = 65,["53"] = 7});
local ____exports = {}
____exports.default = __TS__Class()
local UnitDeathTimeUtil = ____exports.default
UnitDeathTimeUtil.name = "UnitDeathTimeUtil"
function UnitDeathTimeUtil.prototype.____constructor(self)
end
function UnitDeathTimeUtil.setDeathTime(self, deathUnit, deathTime)
    if deathTime == nil then
        deathTime = time
    end
    ____exports.default._sl_deathTimeData[GetHandleId(deathUnit)] = deathTime
end
function UnitDeathTimeUtil.clearDeathTime(self, deathUnit)
    local hid = GetHandleId(deathUnit)
    if ____exports.default._sl_deathTimeData[hid] or ____exports.default._sl_deathTimeData[hid] == 0 then
        ____exports.default._sl_deathTimeData[hid] = nil
    end
end
function UnitDeathTimeUtil.hasDeathTime(self, deathUnit)
    if UnitAlive(deathUnit) then
        return false
    end
    local d = ____exports.default._sl_deathTimeData[GetHandleId(deathUnit)]
    if d == nil then
        return false
    end
    return true
end
function UnitDeathTimeUtil.getDeathTime(self, deathUnit)
    return ____exports.default._sl_deathTimeData[GetHandleId(deathUnit)]
end
function UnitDeathTimeUtil.isTimeOfDeathExceeded(self, deathUnit, timeSlot)
    if timeSlot == nil then
        timeSlot = handleReuseMinTime
    end
    if UnitAlive(deathUnit) then
        return false
    end
    local deathTime = ____exports.default:getDeathTime(deathUnit)
    if deathTime == nil or deathTime == 0 then
        log.errorWithTraceBack("正在对一个已死亡但没有设置DeathTime的单位进行死亡时间判断！请先设置死亡单位的死亡时间")
        return false
    end
    if time - deathTime >= timeSlot * 1000 then
        return true
    end
    return false
end
UnitDeathTimeUtil._sl_deathTimeData = {}
return ____exports
